### PR TITLE
ORC-40: [C++] Support building SearchArgument

### PR DESCRIPTION
**Pull request by wgtmac**
1. Implement Literal class to handle literal values in a predicate of different data types.
2. Port TruthValue, ExpressionTree, PredicateLeaf, and SearchArgument from Java hive-storage-api.
3. Port SargsApplier from Java ORC reader.

### CODE REVIEW

**Comment by wgtmac** [comment]
@omalley @majetideepak @xndai Can you take a look at this patch when you have time? Most of the new classes are ported from java ORC to support predicate pushdown of C++ reader. Thanks!

File: c++/src/sargs/ExpressionTree.cc [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        return leaves[mLeaf];
+      case Operator::CONSTANT:
+        return mConstant;
+      default:
+        throw std::invalid_argument("Unknown operator!");
+    }
+  }
+
+  std::string to_string(TruthValue truthValue) {
+    switch (truthValue) {
```
Comment (chunyang-wen) [inline c++/src/sargs/ExpressionTree.cc]: Can we use a map to store the relationship which uses the enum as the key?

File: c++/src/sargs/ExpressionTree.cc [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ExpressionTree.hh"
+
+#include <cassert>
```
Comment (chunyang-wen) [inline c++/src/sargs/ExpressionTree.cc]: It's better to follow the sequence of header style of existing code.
+ system headers
+ third party headers
+ project headers

File: c++/src/sargs/ExpressionTree.cc [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  }
+
+  TruthValue ExpressionTree::evaluate(
+                                 const std::vector<TruthValue>& leaves) const {
+    TruthValue result;
+    switch (mOperator) {
+      case Operator::OR:
+      {
+        result = mChildren.at(0)->evaluate(leaves);
+        for (size_t i = 1; i < mChildren.size(); ++i) {
```
Comment (chunyang-wen) [inline c++/src/sargs/ExpressionTree.cc]: Is it better to introduce for short-circuit evaluation for `AND` and `OR` operators?

File: c++/src/sargs/Literal.cc [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    return *this;
+  }
+
+  std::string Literal::toString() const {
+    if (mIsNull) {
+      return "null";
+    }
+
+    std::ostringstream sstream;
+    std::string str;
```
Comment (chunyang-wen) [inline c++/src/sargs/Literal.cc]: move `std::string str` to the place where it is used.

File: c++/src/sargs/Literal.cc [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    }
+  }
+
+  bool Literal::operator!=(const Literal& r) const {
+    return !(*this == r);
+  }
+
+  int64_t Literal::getLong() const {
+    if (mIsNull) {
+      throw std::logic_error("cannot call getLong for null!");
```
Comment (chunyang-wen) [inline c++/src/sargs/Literal.cc]: Refact those two if-clauses into a function.

**Comment by majetideepak** [comment]
@wgtmac I will take a look by the end of this week.

File: c++/test/TestPredicateLeaf.cc [hunk 6]
```diff
@@ -0,0 +1,827 @@
...
+    PredicateLeaf pred1(
+      PredicateLeaf::Operator::NULL_SAFE_EQUALS,
+      PredicateType::BOOLEAN,
+      "x",
+      Literal(false));
+    EXPECT_EQ(TruthValue::NO,
+              evaluate(pred1, createBooleanStats(10, 10)));
+    EXPECT_EQ(TruthValue::YES_NO,
+              evaluate(pred1, createBooleanStats(10, 0)));
+
```
Comment (stiga-huang) [inline c++/test/TestPredicateLeaf.cc]: It'd be better to also test "x <=> NULL" (x NULL_SAFE_EQUALS to NULL).

File: c++/src/sargs/ExpressionTree.cc (hunk 2 above)
Comment (wgtmac) [reply c++/src/sargs/ExpressionTree.cc]: I don't think we are following this pattern.

File: c++/src/sargs/ExpressionTree.cc (hunk 3 above)
Comment (wgtmac) [reply c++/src/sargs/ExpressionTree.cc]: Good catch, will do.

File: c++/include/orc/sargs/Literal.hh [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+#include "orc/Int128.hh"
+#include "orc/Vector.hh"
+
+namespace orc {
+
+  /**
+   * Possible types for predicates
+   */
+  enum class PredicateType {
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/include/orc/sargs/Literal.hh]: Yes, the concept of predicate stays unchanged across different systems. I will rename PredicateType to PredicateDataType to eliminate the confusion.

File: c++/include/orc/sargs/TruthValue.hh [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ * limitations under the License.
+ */
+
+#ifndef ORC_TRUTHVALUE_HH
+#define ORC_TRUTHVALUE_HH
+
+namespace orc {
+
+  enum class TruthValue {
+    YES, NO, IS_NULL, YES_NULL, NO_NULL, YES_NO, YES_NO_NULL
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/include/orc/sargs/TruthValue.hh]: These values are copied from [hive-storage-api library](https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hadoop/hive/ql/io/sarg/SearchArgument.java#L39) which unfortunately does not provide more comment than here. I will add some comment here later.

File: c++/src/sargs/PredicateLeaf.cc [hunk 9]
```diff
@@ -0,0 +1,698 @@
...
+
+    auto trueCount = stats.bucketstatistics().count(0);
+    auto falseCount = stats.numberofvalues() - trueCount;
+    switch (op) {
+      case PredicateLeaf::Operator::IS_NULL:
+        return hasNull ? TruthValue::YES_NO : TruthValue::NO;
+      case PredicateLeaf::Operator::NULL_SAFE_EQUALS: {
+        if (literals.at(0).getBool()) {
+           return trueCount == 0 ?
+                  TruthValue::NO : TruthValue::YES_NO;
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/src/sargs/PredicateLeaf.cc]: I think you are right, will fix it.

[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/src/sargs/PredicateLeaf.cc]: You are right. The original implementation is too conservative.

File: c++/src/sargs/PredicateLeaf.cc [hunk 10]
```diff
@@ -0,0 +1,698 @@
...
+      return hasNull ? TruthValue::YES_NO_NULL : TruthValue::YES_NO;
+    }
+
+    auto trueCount = stats.bucketstatistics().count(0);
+    auto falseCount = stats.numberofvalues() - trueCount;
+    switch (op) {
+      case PredicateLeaf::Operator::IS_NULL:
+        return hasNull ? TruthValue::YES_NO : TruthValue::NO;
+      case PredicateLeaf::Operator::NULL_SAFE_EQUALS: {
+        if (literals.at(0).getBool()) {
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/src/sargs/PredicateLeaf.cc]: I am not sure if this is a valid case. I suppose NULL will not appear as a literal in an NULL_SAFE_EQUALS expression.

File: c++/src/sargs/PredicateLeaf.cc [hunk 11]
```diff
@@ -0,0 +1,698 @@
...
+                              , mColumnName(colName)
+                              , mLiterals(literals.begin(), literals.end()) {
+    mHashCode = hashCode();
+    validate();
+  }
+
+  void PredicateLeaf::validate() const {
+    switch (mOperator) {
+      case Operator ::IS_NULL:
+        if (mColumnName.empty()) {
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply c++/src/sargs/PredicateLeaf.cc]: It will check mLiterals inside and will miss the check if validate() is moved to the beginning.

File: c++/src/sargs/ExpressionTree.cc (hunk 1 above)
Comment (wgtmac) [reply c++/src/sargs/ExpressionTree.cc]: I don't think map brings any benefit here.

File: c++/test/TestPredicateLeaf.cc (hunk 6 above)
Comment (wgtmac) [reply c++/test/TestPredicateLeaf.cc]: Done. I have resolved all issues. Can you take a look again? Thanks! @stiga-huang

**Comment by stiga-huang** [comment]
LGTM. Thanks for addressing the comments!

**Comment by majetideepak** [comment]
@wgtmac I plan to do a lazy review after merge for this. Since all of this is new code, I am not too worried about regressions. 
Thanks to @stiga-huang for giving a round of feedback.

**Comment by majetideepak** [comment]
Feel free to merge this if its ready.
